src/pybind/mgr/orchestrator.py

```diff
@@ -0,0 +1,266 @@
+
+"""
+ceph-mgr orchestrator interface
+
+This is a DRAFT for discussion.
+
+Goal: enable UI workflows for cluster service management
+      (such as creating OSDs, in addition to stateless services)
+      using common concepts that are implemented by
+      diverse backends such as Rook, DeepSea, ceph-ansible
+
+Concepts:
+    "Stateful service": a daemon that uses local storage, such as OSD or mon.
+    "Stateless service": a daemon that doesn't use any local storage, such
+                         as an MDS, RGW, nfs-ganesha, iSCSI gateway.
+    "Label": arbitrary string tags that may be applied by administrators
+             to nodes.  Typically administrators use labels to indicate
+             which nodes
+
+Constraints:
+    1. The orchestrator is to be the source of truth for
+       all the physical information, and will be queried directly
+       as needed (i.e. no in-Ceph database of hardware etc).
+    2. The orchestrator handles placement of collections of stateless
+       services.
+    3. The orchestrator accepts explicit placement of individual stateful
+       services, and optionally also accepts label-based automatic placement.
+       (i.e. it *must* support "create OSD at host1:/dev/sdb", and it *may*
+        support "create OSDs on hosts with label=ceph-osd")
+    4. Ceph is not responsible for installing anything on nodes, or hooking
+       up hosts to the orchestrator: nodes and drives only exist to Ceph
+       once they're exposed by the orchestrator.
+
+Flexible features:
+    1. Orchestrator may be aware of one or more "special features" for
+       block devices, such as block-level encryption.  These special
+       features may be reported during discovery or enabled during
+       OSD creation.
+    2. Orchestrator may not have a concept of labels: if it doesn't,
+       any label-filtered calls will match everything.
+
+Excluded functionality:
+    1. No support for multipathed drives: all block devices are to be
+    reported from one host only.
+    2. No networking inventory or configuration.
+
+This is a DRAFT for discussion.
+"""
+
+
+class Orchestrator(object):
+    """
+    All calls in this class are blocking on network IO and may
+    take noticeable length of time to run.  Wrap calls appropriately
+    in long-running task infrastructure when calling from interactive
+    code.
+    """
+
+    # >>> From 21 March meeting
+    def get_inventory(self, node_filter=None):
+        # Return list of InventoryHost
+        raise NotImplementedError()
+
+    def add_stateful_service(self, service_type, spec):
+        """
+        service_type: one of mon, osd
```

Comment:
We should do input sanitization on this since the inputs are known and limited. I.e.,
```py
assert service_type in ['mon', 'osd']
```